ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/MerkleTree.java

```
@@ -38,45 +38,110 @@ protected MerkleTree(int treeDepth) {
     zeroHashes = generateZeroHashes(treeDepth);
   }
 
-  abstract void add(Bytes32 leaf);
+  public abstract void add(Bytes32 leaf);
 
-  protected abstract int getNumberOfLeaves();
+  public abstract int getNumberOfLeaves();
 
   protected static List<Bytes32> generateZeroHashes(int height) {
     List<Bytes32> zeroHashes = new ArrayList<>();
-    for (int i = 0; i < height; i++) {
-      zeroHashes.add(Bytes32.ZERO);
-    }
-    for (int i = 0; i < height - 1; i++) {
-      zeroHashes.set(i + 1, Hash.sha2_256(Bytes.concatenate(zeroHashes.get(i), zeroHashes.get(i))));
+    zeroHashes.add(Bytes32.ZERO);
+    for (int i = 1; i < height; i++) {
+      zeroHashes.add(
+          i, Hash.sha2_256(Bytes.concatenate(zeroHashes.get(i - 1), zeroHashes.get(i - 1))));
     }
     return zeroHashes;
   }
 
-  public SSZVector<Bytes32> getProofTreeByValue(Bytes32 value) {
+  public SSZVector<Bytes32> getProof(Bytes32 value) {
     int index = tree.get(0).indexOf(value);
```

We should probably throw an exception of the value isn't found.

---

Done.